rascal-lsp/src/main/rascal/lang/rascal/lsp/Rename.rsc

```diff
@@ -0,0 +1,272 @@
+    | captures(rel[loc use, loc oldDef, loc newDef] captures)
+    ;
+
+data RenameException
+    = illegalRename(loc location, set[IllegalRenameReason] reason)
+    | unsupportedRename(rel[loc location, str message] issues)
+    | unexpectedFailure(str message)
+;
+
+private set[loc] getUses(TModel tm, loc def) {
+    return invert(getUseDef(tm))[def];
+}
+
+private set[Define] getDefines(TModel tm, loc useDef) {
+    set[loc] defLocs = tm.useDef[useDef] != {}
+                     ? tm.useDef[useDef] // `useDef` is a use location
+                     : {useDef};         // `useDef` is a definition location
+
+    return {tm.definitions[d] | d <- defLocs, tm.definitions[d]?};
+}
+
+void throwIfNotEmpty(&T(&A) R, &A arg) {
+    if (arg != {}) {
+        throw R(arg);
+    }
+}
+
+set[IllegalRenameReason] checkLegalName(str name) {
+    try {
+        parse(#Name, escapeName(name));
+        return {};
+    } catch ParseError(_): {
+        return {invalidName(name)};
+    }
+}
+
+set[IllegalRenameReason] checkCausesDoubleDeclarations(set[Define] currentDefs, set[Define] newDefs) {
+    // Is newName already resolvable from a scope where <current-name> is currently declared?
+    rel[loc new, loc old] doubleDeclarations = {<nD.defined, cD.defined> | Define nD <- newDefs
+                                                               , Define cD <- currentDefs
+                                                               , isContainedIn(cD.defined, nD.scope)
+    };
+
+    if (doubleDeclarations != {}) {
+        return {doubleDeclaration(doubleDeclarations)};
+    }
+
+    return {};
+}
+
+set[Define] findImplicitDefinitions(TModel tm, start[Module] m, set[Define] newDefs) {
+    set[loc] maybeImplicitDefs = {l | /QualifiedName n := m, just(l) := locationOfName(n)};
+    return {def | Define def <- newDefs, (def.idRole is variableId && def.defined in tm.useDef<0>) || (def.idRole is patternVariableId && def.defined in maybeImplicitDefs)};
+}
+
+set[IllegalRenameReason] checkCausesCaptures(TModel tm, start[Module] m, set[Define] currentDefs, set[loc] currentUses, set[Define] newDefs) {
+    set[Define] newNameImplicitDefs = findImplicitDefinitions(tm, m, newDefs);
+    set[Define] newNameExplicitDefs = newDefs - newNameImplicitDefs;
+
+    // Will this rename turn an implicit declaration of `newName` into a use of a current declaration?
+    rel[loc use, loc oldDef, loc newDef] implicitDeclBecomesUseOfCurrentDecl =
+        {<nD.defined, nD.defined, cD> | Define nD <- newNameImplicitDefs
+                                      , <loc cS, loc cD> <- currentDefs<scope, defined>
+                                      , isContainedIn(nD.defined, cS)
+        };
+
+    // Will this rename turn a use of <oldName> into a use of an existing declaration of <newName> (shadowing)?
+    rel[loc use, loc oldDef, loc newDef] explicitDeclShadowsCurrentDecl =
+        {<cU, cD.defined, nD.defined> | Define nD <- newNameExplicitDefs
```

Comment:
if this is a efficiency bottleneck then rewriting a bit with the `o` composition operator may take off the edge. It typically avoids one more iteration through the target set, so takes off one complexity order. I think there are two hidden `o` operators here before two filters.

Reply:
I will think about that. Performance has not been my primary concern and since this should only consider local variables (otherwise an exception will be thrown before this code is even called), I am not too worried right now.

Reply:
Actually went ahead and rewrote this. Much better, thanks!